zarr/storage.py

```diff
@@ -946,6 +946,87 @@ def atexit_rmglob(path,
             rmtree(p)
 
 
+class FSStore(MutableMapping):
+
+    def __init__(self, url, normalize_keys=True, key_separator='.',
+                 **storage_options):
+        import fsspec
+        self.path = url
+        self.normalize_keys = normalize_keys
+        self.key_separator = key_separator
+        self.map = fsspec.get_mapper(url, **storage_options)
+        self.fs = self.map.fs  # for direct operations
+
+    def _normalize_key(self, key):
+        key = normalize_storage_path(key)
+        if key:
+            *bits, end = key.split('/')
+            key = '/'.join(bits + [end.replace('.', self.key_separator)])
+        return key.lower() if self.normalize_keys else key
+
+    def __getitem__(self, key):
+        key = self._normalize_key(key)
+        return self.map[key]
+
+    def __setitem__(self, key, value):
+        key = self._normalize_key(key)
+        path = self.dir_path(key)
+        value = ensure_contiguous_ndarray(value)
+        try:
+            if self.fs.isdir(path):
+                self.fs.rm(path, recursive=True)
+            self.map[key] = value
+        except IOError:
+            raise KeyError(key)
+
+    def __delitem__(self, key):
+        key = self._normalize_key(key)
+        path = self.dir_path(key)
+        if self.fs.isdir(path):
+            self.fs.rm(path, recursive=True)
+        else:
+            del self.map[key]
+
+    def __contains__(self, key):
+        key = self._normalize_key(key)
+        return key in self.map
+
+    def __eq__(self, other):
+        return type(self) == type(other) and self.map == other.map
+
+    def keys(self):
+        return iter(self.map)
+
+    def __iter__(self):
+        return self.keys()
+
+    def __len__(self):
+        return len(list(self.keys()))
+
+    def dir_path(self, path=None):
+        store_path = normalize_storage_path(path)
+        return self.map._key_to_str(store_path)
+
+    def listdir(self, path=None):
+        dir_path = self.dir_path(path)
+        try:
+            return sorted(p.rsplit('/', 1)[-1] for p in self.fs.ls(dir_path))
```

Comment:
This line throws an AttributeError for `group.arrays()` when serving a directory created with:

```
import zarr
store = zarr.DirectoryStore('dir.zarr')
group = zarr.group(store=store, overwrite=True)
z = group.zeros("test", shape=(4, 4))
z[:] = 4
```

via `python -m http.server`:

```
In [3]: from zarr.storage import FSStore

In [4]: store = FSStore("http://localhost:8000")

In [5]: group = zarr.group(store=store)

In [6]: group.arrays()
Out[6]: <generator object Group._array_iter at 0x7fd2e0bff258>

In [7]: list(group.arrays())
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-7-b56aa2c546df> in <module>
----> 1 list(group.arrays())

~/opt/zarr/zarr/hierarchy.py in _array_iter(self, keys_only, method, recurse)
    477
    478     def _array_iter(self, keys_only, method, recurse):
--> 479         for key in sorted(listdir(self._store, self._path)):
    480             path = self._key_prefix + key
    481             if contains_array(self._store, path):

~/opt/zarr/zarr/storage.py in listdir(store, path)
    165     if hasattr(store, 'listdir'):
    166         # pass through
--> 167         return store.listdir(path)
    168     else:
    169         # slow version, iterate through all keys

~/opt/zarr/zarr/storage.py in listdir(self, path)
   1011         dir_path = self.dir_path(path)
   1012         try:
-> 1013             return sorted(p.rsplit('/', 1)[-1] for p in self.fs.ls(dir_path))
   1014         except IOError:
   1015             return []

~/opt/zarr/zarr/storage.py in <genexpr>(.0)
   1011         dir_path = self.dir_path(path)
   1012         try:
-> 1013             return sorted(p.rsplit('/', 1)[-1] for p in self.fs.ls(dir_path))
   1014         except IOError:
   1015             return []

AttributeError: 'dict' object has no attribute 'rsplit'
```

while the current direct use of `from fsspec.implementations.http import HTTPFileSystem` shows:

```
In [33]: list(zg.arrays())
Out[33]: [('test', <zarr.core.Array '/test' (4, 4) float64 read-only>)]
```

Reply:
fixed